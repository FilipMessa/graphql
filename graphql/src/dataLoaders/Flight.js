// @flow

import DataLoader from 'dataloader';
import dateFns from 'date-fns';
import _ from 'lodash';

import { get } from '../services/HttpRequest';
import config from '../../config/application';
import LocationDataLoader from '../dataLoaders/Location';
import localeMap from '../inputs/LocaleMap';

import type { Flight } from '../types/Flight';
import type { LocationVariants, RadiusLocation } from '../types/Location';

type QueryParameters = {|
  from: Array<LocationVariants>,
  to: Array<LocationVariants>,
  dateFrom: null | Date,
  dateTo: null | Date,
  daysInDestinationFrom: null | number,
  daysInDestinationTo: null | number,
  currency: null | string,
  adults: null | number,
  locale: null | string,
|};

type BackendAPIResponse = {|
  currency: string,
  data: Flight[],
|};

export default class FlightDataloader {
  flightDataLoader: DataLoader<QueryParameters, BackendAPIResponse>;
  locationDataLoader: LocationDataLoader;

  constructor(locationDataLoader: LocationDataLoader) {
    this.flightDataLoader = new DataLoader(
      (searchParameters: QueryParameters[]) => {
        return this.batchGetLocations(searchParameters);
      },
      {
        cacheKeyFn: key => {
          return JSON.stringify(key);
        },
      },
    );
    this.locationDataLoader = locationDataLoader;
  }

  async load(searchParameters: QueryParameters): Promise<BackendAPIResponse> {
    return this.flightDataLoader.load(searchParameters);
  }

  async loadMany(
    searchParameters: QueryParameters[],
  ): Promise<BackendAPIResponse[]> {
    return this.flightDataLoader.loadMany(searchParameters);
  }

  async batchGetLocations(
    searchParameters: QueryParameters[],
  ): Promise<Array<BackendAPIResponse | Error>> {
    return await Promise.all(
      searchParameters.map(searchParameters =>
        this._fetchFlights(searchParameters),
      ),
    );
  }

  async _fetchFlights(searchParameters: QueryParameters) {
    const {
      from,
      to,
      dateFrom,
      dateTo,
      daysInDestinationFrom,
      daysInDestinationTo,
      currency,
      adults,
      locale,
    } = searchParameters;
    // see: https://github.com/tc39/proposal-object-rest-spread/issues/45
    const parameters = {
      flyFrom: this._normalizeLocations(from),
      to: this._normalizeLocations(to),
      ...(dateFrom && {
        dateFrom: dateFns.format(dateFrom, 'DD/MM/YYYY'),
      }),
      ...(dateTo && {
        dateTo: dateFns.format(dateTo, 'DD/MM/YYYY'),
      }),
    };
    if (daysInDestinationFrom !== null) {
      parameters.daysInDestinationFrom = daysInDestinationFrom;
    }
    if (daysInDestinationTo !== null) {
      parameters.daysInDestinationTo = daysInDestinationTo;
    }
    if (currency !== null) {
      parameters.curr = currency;
    }
    if (adults !== null) {
      parameters.adults = adults;
    }
    if (locale !== null) {
      parameters.locale = localeMap[locale];
    }

    const firstTry = await get(config.restApiEndpoint.allFlights(parameters));
    if (firstTry.data.length > 0) {
      return firstTry;
    }

    const [flyFrom, flyTo] = await Promise.all([
      this._normalizeLocationsUsingRefetch(from, locale),
      this._normalizeLocationsUsingRefetch(to, locale),
    ]);
    parameters.flyFrom = flyFrom;
    parameters.to = flyTo;
    return get(config.restApiEndpoint.allFlights(parameters));
  }

  _normalizeLocations(locations: LocationVariants[]): string[] {
    return locations.map(location => {
      if (location.location !== undefined) {
        return location.location;
      } else {
        return this._stringifyRadius(location);
      }
    });
  }

  async _normalizeLocationsUsingRefetch(
    locations: LocationVariants[],
    locale: null | string,
  ) {
    const options = locale !== null ? { locale } : {};
    const normalizedLocations = locations.map(location => {
      if (location.location === undefined) {
        return Promise.resolve(this._stringifyRadius(location));
      }
      // adjust exact location using backend location API
      return this.locationDataLoader
        .load(location.location, options)
        .then(newLocation => newLocation.locationId); // eslint-disable-line promise/prefer-await-to-then
    });
    return Promise.all(normalizedLocations);
  }

  _stringifyRadius({ radius }: RadiusLocation): string {
    const lat = _.round(radius.lat, 2);
    const lng = _.round(radius.lng, 2);
    const distance = _.round(radius.radius, 0);
    return `${lat}-${lng}-${distance}km`;
  }
}
