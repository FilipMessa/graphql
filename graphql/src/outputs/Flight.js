// @flow

import _ from 'lodash';
import { GraphQLInt, GraphQLList, GraphQLObjectType } from 'graphql';
import GraphQLRouteStop from './RouteStop';
import GraphQLLeg from './Leg';
import GraphQLAirline from './Airline';
import GraphQLPrice from './Price';
import type { GraphqlContextType } from '../services/GraphqlContext';
import FlightDurationInMinutes from '../resolvers/FlightDuration';

import type { Price } from '../types/Price';
import type { DepartureArrival, Flight, Leg, Airline } from '../types/Flight';

export default new GraphQLObjectType({
  name: 'Flight',
  fields: {
    airlines: {
      type: new GraphQLList(GraphQLAirline),
      description: 'List of all Airlines involved.',
      resolve: async (
        { airlines }: Flight,
        args: Object,
        { dataLoader }: GraphqlContextType,
      ): Promise<Array<?Airline>> =>
        _.uniq(airlines).map(airlineCode =>
          dataLoader.airline.load(airlineCode),
        ),
    },

    arrival: {
      type: GraphQLRouteStop,
      resolve: ({ arrival }: Flight): DepartureArrival => arrival,
    },

    departure: {
      type: GraphQLRouteStop,
      resolve: ({ departure }: Flight): DepartureArrival => departure,
    },

    duration: {
      type: GraphQLInt,
      description: 'Flight duration in minutes.',
      resolve: ({ departure, arrival }: Flight): ?number =>
        FlightDurationInMinutes(departure, arrival),
    },

    legs: {
      type: new GraphQLList(GraphQLLeg),
      description: 'Flight segments, e.g. stopover, change of aircraft, etc.',
      resolve: ({ legs }: Flight): Array<Leg> => legs,
    },

    price: {
      type: GraphQLPrice,
      description: 'Total flight price.',
      resolve: ({ price }: Flight): Price => price,
    },
  },
});
