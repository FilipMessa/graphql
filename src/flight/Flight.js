// @flow

import type { Price } from '../common/types/Price';

export type DepartureArrival = {|
  when: ?{|
    // null if doesn't exist
    utc: Date, // UTC +0
    local: Date, // UTC != 0
  |},
  where: {|
    code: string,
    cityName: string,
    cityId: string,
  |},
|};

export type Flight = {|
  id: string,
  airlines: Array<string>,
  arrival: DepartureArrival,
  departure: DepartureArrival,
  legs: Array<Leg>,
  price: Price,
  bookingToken: string,
  passengers: number,
|};

export type Leg = {|
  id: string,
  recheckRequired: boolean,
  isReturn: boolean,
  flightNo: number,
  departure: DepartureArrival,
  arrival: DepartureArrival,
  airlineCode: string,
|};

export type Airline = {|
  name: string,
  logoUrl: string,
  code: string,
  isLowCost: boolean,
|};
