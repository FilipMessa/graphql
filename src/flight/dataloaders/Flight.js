// @flow

import DataLoader from 'dataloader';
import { DateTime } from 'luxon';

import { get } from '../../common/services/HttpRequest';
import config from '../../../config/application';
import LocationDataLoader from '../../location/dataloaders/Location';
import LocationsDataLoader from '../../location/dataloaders/Locations';
import LocaleMap from '../../common/types/enums/LocaleMap';

import type { Flight } from '../Flight';
import type { LocationVariants, RadiusLocation } from '../../location/Location';

type Duration = {|
  maxFlightDuration: null | number,
  stopovers: null | { from: ?number, to: ?number },
|};

type Filters = {|
  maxStopovers: null | number,
  duration: null | Duration,
|};

type QueryParameters = {|
  from: Array<LocationVariants>,
  to: Array<LocationVariants>,
  dateFrom: Date,
  dateTo: Date,
  returnFrom: null | Date,
  returnTo: null | Date,
  typeFlight: 'return' | 'oneway',
  currency: null | string,
  adults: null | number,
  locale: null | string,
  filters: null | Filters,
|};

type BackendAPIResponse = {|
  currency: string,
  data: Flight[],
|};

export default class FlightDataloader {
  flightDataLoader: DataLoader<QueryParameters, BackendAPIResponse>;
  locationDataLoader: LocationDataLoader;
  locationsDataLoader: LocationsDataLoader;

  constructor(
    locationDataLoader: LocationDataLoader,
    locationsDataLoader: LocationsDataLoader,
  ) {
    this.locationsDataLoader = locationsDataLoader;
    this.flightDataLoader = new DataLoader(
      (searchParameters: $ReadOnlyArray<QueryParameters>) => {
        return this.batchGetLocations(searchParameters);
      },
      {
        cacheKeyFn: key => {
          return JSON.stringify(key);
        },
      },
    );
    this.locationDataLoader = locationDataLoader;
  }

  async load(searchParameters: QueryParameters): Promise<BackendAPIResponse> {
    return this.flightDataLoader.load(searchParameters);
  }

  async loadMany(
    searchParameters: QueryParameters[],
  ): Promise<BackendAPIResponse[]> {
    return this.flightDataLoader.loadMany(searchParameters);
  }

  async batchGetLocations(
    searchParameters: $ReadOnlyArray<QueryParameters>,
  ): Promise<Array<BackendAPIResponse>> {
    return await Promise.all(
      searchParameters.map(searchParameters =>
        this._fetchFlights(searchParameters),
      ),
    );
  }

  async _fetchFlights(searchParameters: QueryParameters) {
    const {
      from,
      to,
      dateFrom,
      dateTo,
      returnFrom,
      returnTo,
      typeFlight,
      currency,
      adults,
      locale,
      filters,
    } = searchParameters;

    const parameters: Object = {
      flyFrom: this._normalizeLocations(from),
      to: this._normalizeLocations(to),
      dateFrom: this._formatDateForApi(dateFrom),
      dateTo: this._formatDateForApi(dateTo),
      ...(returnFrom && {
        returnFrom: this._formatDateForApi(returnFrom),
      }),
      ...(returnTo && {
        returnTo: this._formatDateForApi(returnTo),
      }),
      ...(typeFlight && {
        typeFlight: typeFlight,
      }),
    };
    if (currency !== null) {
      parameters.curr = currency;
    }
    if (adults !== null) {
      parameters.adults = adults;
    }
    if (locale !== null) {
      parameters.locale = LocaleMap[locale];
    }
    if (filters) {
      if (filters.maxStopovers !== null) {
        parameters.maxStopovers = filters.maxStopovers;
      }
      if (filters.duration) {
        const duration = filters.duration;
        if (duration.maxFlightDuration !== null) {
          parameters.maxFlyDuration = duration.maxFlightDuration;
        }
        if (duration.stopovers && duration.stopovers.from !== null) {
          parameters.stopoverfrom = duration.stopovers.from;
        }
        if (duration.stopovers && duration.stopovers.to !== null) {
          parameters.stopoverto = duration.stopovers.to;
        }
      }
    }

    const firstTry = await get(config.restApiEndpoint.allFlights(parameters));
    if (firstTry.data.length > 0) {
      return firstTry;
    }

    const [flyFrom, flyTo] = await Promise.all([
      this._normalizeLocationsUsingRefetch(from, locale),
      this._normalizeLocationsUsingRefetch(to, locale),
    ]);
    parameters.flyFrom = flyFrom;
    parameters.to = flyTo;
    return get(config.restApiEndpoint.allFlights(parameters));
  }

  /**
   * Input dates should be in UTC. This method changes the date format
   * to be compatible with underlying API. It keeps the same timezone (UTC)
   * and ignores local timezone. So no matter where you run this code
   * it should always work as expected (returns the same date in different format).
   */
  _formatDateForApi = (date: Date): string =>
    DateTime.fromJSDate(date, { zone: 'utc' }).toFormat(
      'dd/LL/yyyy', // http://moment.github.io/luxon/docs/manual/usage/formatting.html
    );

  _normalizeLocations(locations: LocationVariants[]): string[] {
    return locations.map(location => {
      if (location.location !== undefined) {
        return location.location;
      } else {
        return this._stringifyRadius(location);
      }
    });
  }

  async _normalizeLocationsUsingRefetch(
    locations: LocationVariants[],
    locale: null | string,
  ) {
    const options = locale !== null ? { locale } : {};
    const normalizedLocations = locations.map(location => {
      if (location.location === undefined) {
        return Promise.resolve(this._stringifyRadius(location));
      }
      // adjust exact location using backend location API
      return this.locationsDataLoader
        .loadByTerm(location.location, options)
        .then(possibleValues => possibleValues[0].locationId); // eslint-disable-line promise/prefer-await-to-then
    });
    return Promise.all(normalizedLocations);
  }

  _stringifyRadius({ radius }: RadiusLocation): string {
    const lat = Math.round(radius.lat * 100) / 100;
    const lng = Math.round(radius.lng * 100) / 100;
    const distance = Math.round(radius.radius);
    return `${lat}-${lng}-${distance}km`;
  }
}
