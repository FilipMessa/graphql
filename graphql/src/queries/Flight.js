import { GraphQLNonNull } from 'graphql';
import Flight from './../types/Flight';

export default {
  type: new GraphQLNonNull(Flight),
  resolve() {
    return {
      arrival: {
        id: 'arrival-uuid', // FIXME: connect to the API
      },
      departure: {
        id: 'departure-uuid', // FIXME: connect to the API
      },
    };
  },
};
