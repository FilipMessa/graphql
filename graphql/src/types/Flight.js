// @flow

import _ from 'lodash';
import {
  GraphQLObjectType,
  GraphQLNonNull,
  GraphQLList,
  GraphQLInt,
} from 'graphql';
import GraphQLRouteStop from './RouteStop';
import GraphQLLeg from './Leg';
import GraphQLAirline from './Airline';
import { createAirline } from '../dataLoaders/Airline';
import { flightDurationInMinutes } from '../services/GraphqlResolvers';

import type {
  AirlineType,
  ArrivalType,
  FlightType,
  DepartureType,
  LegType,
} from '../Entities';

export default new GraphQLObjectType({
  name: 'Flight',
  fields: {
    airlines: {
      type: new GraphQLList(GraphQLAirline),
      resolve: ({ airlines }: FlightType): Array<AirlineType> =>
        _.uniq(airlines).map(airlineCode => createAirline(airlineCode)),
    },

    arrival: {
      type: new GraphQLNonNull(GraphQLRouteStop),
      resolve: ({ arrival }: FlightType): ArrivalType => arrival,
    },

    departure: {
      type: new GraphQLNonNull(GraphQLRouteStop),
      resolve: ({ departure }: FlightType): DepartureType => departure,
    },

    duration: {
      type: GraphQLInt,
      resolve: ({ departure, arrival }: FlightType): ?number =>
        flightDurationInMinutes(departure, arrival),
    },

    legs: {
      type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(GraphQLLeg))),
      resolve: ({ legs }: FlightType): Array<LegType> => legs,
    },
  },
});
