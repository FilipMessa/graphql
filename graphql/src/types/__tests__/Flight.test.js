// @flow

import Flight from '../Flight';

const fields = Flight.getFields();

describe('arrival field', () => {
  it('should be non-null Arrival type', () => {
    expect(fields.arrival.type.toString()).toBe('Arrival!');
  });
});

describe('departure field', () => {
  it('should be non-null Departure string', () => {
    expect(fields.departure.type.toString()).toBe('Departure!');
  });
});
