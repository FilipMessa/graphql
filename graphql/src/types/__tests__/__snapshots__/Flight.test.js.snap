// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flight type should have valid fields 1`] = `
Object {
  "airlines": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "airlines",
    "resolve": [Function],
    "type": "[Airline]",
  },
  "arrival": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "arrival",
    "resolve": [Function],
    "type": "RouteStop!",
  },
  "departure": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "departure",
    "resolve": [Function],
    "type": "RouteStop!",
  },
  "duration": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "duration",
    "resolve": [Function],
    "type": "Int",
  },
  "legs": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "legs",
    "resolve": [Function],
    "type": "[Leg!]!",
  },
}
`;
